source/_posts/2017-11-02-kafka-streams-dsl-vs-processor-api.markdown

```diff
@@ -0,0 +1,537 @@
+because business logic can be expressed in a few lines of code.
+All stateless and stateful transformations are defined using declarative, 
+functional programming style (filter, map, flatMap, reduce, aggregate operations).
+Kafka Stream DSL encapsulates most of the stream processing complexity
+but unfortunately it also hides many useful knobs and switches. 
+
+* [Kafka Processor API](https://docs.confluent.io/current/streams/developer-guide.html#processor-api) 
+provides a low level, imperative way to define stream processing logic.
+At first sight Processor API could look hostile but finally gives much more flexibility to developer.
+This blog post shows that hand-crafted stream processors might be a magnitude more efficient than
+a naive implementation using Kafka DSL.
+
+* [KSQL](https://www.confluent.io/product/ksql/) 
+is a promise that stream processing could be expressed by anyone using SQL as the language.
+It offers an easy way to express stream processing transformations as an alternative to writing 
+an application in a programming language such as Java.
+Moreover, processing transformation written in SQL like language can be highly optimized
+by execution engine without any developer effort. 
+KSQL was released recently and it is still at very early development stage.
+
+In the first part of this blog post I'll define simple but still realistic business problem to solve.
+Then you will learn how to implement this use case with Kafka Stream DSL 
+and how much the processing performance is affected by this naive solution.
+At this moment you could stop reading and scale-up Kafka cluster ten times to fulfill business requirements 
+or you could continue reading and learn how to optimize the processing with low level Kafka Processor API.
+
+## Business Use Case
+
+Let's imagine web based e-commerce platform with fabulous recommendation and advertisement subsystems.
+Every client during visit gets personalized recommendations and advertisements,
+the conversion is extraordinary high and platform earns additional profits from advertisers.
+To build comprehensive recommendation models, 
+such system needs to know everything about clients traits and their behaviour.
+
+To make it possible, e-commerce platform reports all clients activities as unbounded stream 
+of page views and events.
+Every time, the client enters web page, so-called page view is sent to Kafka cluster. 
+Page view defines web page attributes like request URI, referrer URI, user agent, active A/B experiments
+and many more.
+In addition to page view all important actions are reported as events, e.g: search, add to cart or checkout.
+To get complete view of the activity stream, collected events need to be enriched by data from page views.
+
+## Data Model
+
+Because most of the processing logic is built within context of given client, 
+page views and events are evenly partitioned on Kafka topics by client identifier.
```

Comment:
by the

Reply:
zastanawiam się czy nie lepiej czytałoby się tych dwóch zdań w odwrotnym szyku
i może with regard to a zamiast within context of?